// nuxt-laravel-echo@0.2.7/dist/module.mjs
import { addTypeTemplate, defineNuxtModule, createResolver, useLogger, addPlugin, addImportsDir } from '@nuxt/kit';
import { defu } from 'defu';

const registerTypeTemplates = (resolver) => {
  addTypeTemplate({
    filename: "types/echo.d.ts",
    getContents: () => `// Generated by nuxt-laravel-echo module
import type { EchoAppConfig } from '${resolver.resolve("./runtime/types/config.ts")}';

declare module 'nuxt/schema' {
    interface AppConfig {
        echo?: EchoAppConfig;
    }
    interface AppConfigInput {
        echo?: EchoAppConfig;
    }
}

declare module '@nuxt/schema' {
    interface AppConfig {
        echo?: EchoAppConfig;
    }
    interface AppConfigInput {
        echo?: EchoAppConfig;
    }
}

export {};`
  });
};

const MODULE_NAME = "nuxt-laravel-echo";
const defaultModuleOptions = {
  broadcaster: "reverb",
  host: "localhost",
  port: 8080,
  scheme: "https",
  transports: ["ws", "wss"],
  authentication: {
    mode: "cookie",
    baseUrl: "http://localhost:80",
    authEndpoint: "/broadcasting/auth",
    csrfEndpoint: "/sanctum/csrf-cookie",
    csrfCookie: "XSRF-TOKEN",
    csrfHeader: "X-XSRF-TOKEN"
  },
  logLevel: 3
};
const module = defineNuxtModule({
  meta: {
    name: MODULE_NAME,
    configKey: "echo",
    compatibility: {
      nuxt: ">=3.13.0"
    }
  },
  defaults: defaultModuleOptions,
  setup(_options, _nuxt) {
    const resolver = createResolver(import.meta.url);
    const logger = useLogger(MODULE_NAME, { level: _options.logLevel });
    _nuxt.options.build.transpile.push(resolver.resolve("./runtime"));
    _nuxt.options.runtimeConfig.public.echo = defu(
      _nuxt.options.runtimeConfig.public.echo,
      _options
    );
    addPlugin(resolver.resolve("./runtime/plugin.client"));
    addImportsDir(resolver.resolve("./runtime/composables"));
    registerTypeTemplates(resolver);
    logger.info("Laravel Echo module initialized!");
  }
});

export { module as default };
