// nuxt-auth-sanctum@1.2.0/dist/module.mjs
import { addTypeTemplate, defineNuxtModule, createResolver, useLogger, addPlugin, addImportsDir, addRouteMiddleware, addServerHandler } from '@nuxt/kit';
import { defu } from 'defu';
import { relative, resolve } from 'pathe';

const defaultModuleOptions = {
  baseUrl: "http://localhost:80",
  mode: "cookie",
  userStateKey: "sanctum.user.identity",
  redirectIfAuthenticated: false,
  redirectIfUnauthenticated: false,
  endpoints: {
    csrf: "/sanctum/csrf-cookie",
    login: "/login",
    logout: "/logout",
    user: "/api/user"
  },
  csrf: {
    cookie: "XSRF-TOKEN",
    header: "X-XSRF-TOKEN"
  },
  client: {
    retry: false,
    initialRequest: true
  },
  redirect: {
    keepRequestedRoute: false,
    onLogin: "/",
    onLogout: "/",
    onAuthOnly: "/login",
    onGuestOnly: "/"
  },
  globalMiddleware: {
    enabled: false,
    prepend: false,
    allow404WithoutAuth: true
  },
  logLevel: 3,
  appendPlugin: false,
  serverProxy: {
    enabled: false,
    route: "/api/sanctum",
    baseUrl: "http://localhost:80"
  }
};

const registerTypeTemplates = (resolver) => {
  addTypeTemplate({
    filename: "types/sanctum.d.ts",
    getContents: ({ nuxt }) => {
      const { buildDir } = nuxt.options;
      const getRelativePath = (path) => relative(resolve(buildDir, "./types"), resolver.resolve(path));
      return `// Generated by nuxt-auth-sanctum module
import type { NuxtApp } from '#app'
import type { HookResult } from '@nuxt/schema'
import type { FetchContext } from 'ofetch'
import type { ConsolaInstance } from 'consola'
import type { SanctumAppConfig } from '${getRelativePath("./runtime/types/config.ts")}';
import type { SanctumGlobalMiddlewarePageMeta } from '${getRelativePath("./runtime/types/meta.ts")}';

declare module 'nuxt/schema' {
  interface AppConfig {
    sanctum?: SanctumAppConfig
  }
  interface AppConfigInput {
    sanctum?: SanctumAppConfig
  }
}

declare module '@nuxt/schema' {
  interface AppConfig {
    sanctum?: SanctumAppConfig
  }
  interface AppConfigInput {
    sanctum?: SanctumAppConfig
  }
}

declare module '#app' {
  interface PageMeta {
    /**
     * Sanctum global middleware page configuration.
     */
    sanctum?: Partial<SanctumGlobalMiddlewarePageMeta>
  }

  interface RuntimeNuxtHooks {
    /**
     * Triggers when receiving an error response.
     */
    'sanctum:error:response': (context: FetchContext) => HookResult
    /**
     * Triggers when receiving an error on fetch request.
     */
    'sanctum:error:request': (context: FetchContext) => HookResult
    /**
     * Triggers when the user has been redirected.
     */
    'sanctum:redirect': (path: string) => HookResult
    /**
     * Triggers when an initial user identity request has been made.
     */
    'sanctum:init': () => HookResult
    /**
     * Triggers when user identity has been refreshed.
     */
    'sanctum:refresh': () => HookResult
    /**
     * Triggers when user successfully logs in.
     */
    'sanctum:login': () => HookResult
    /**
     * Triggers when user successfully logs out.
     */
    'sanctum:logout': () => HookResult
    /**
     * Triggers on every client request.
     */
    'sanctum:request': (app: NuxtApp, ctx: FetchContext, logger: ConsolaInstance) => HookResult
    /**
     * Triggers on every server response.
     */
    'sanctum:response': (app: NuxtApp, ctx: FetchContext, logger: ConsolaInstance) => HookResult
  }
}

export {}`;
    }
  });
};

const MODULE_NAME = "nuxt-auth-sanctum";
const module = defineNuxtModule({
  meta: {
    name: MODULE_NAME,
    configKey: "sanctum",
    compatibility: {
      nuxt: ">=3.17.0"
    }
  },
  defaults: defaultModuleOptions,
  setup(_options, _nuxt) {
    const resolver = createResolver(import.meta.url);
    const sanctumConfig = defu(
      _nuxt.options.runtimeConfig.public.sanctum,
      _options
    );
    _nuxt.options.build.transpile.push(resolver.resolve("./runtime"));
    _nuxt.options.runtimeConfig.sanctum = sanctumConfig;
    const publicSanctumConfig = { ...sanctumConfig };
    delete publicSanctumConfig.serverProxy;
    _nuxt.options.runtimeConfig.public.sanctum = publicSanctumConfig;
    const logger = useLogger(MODULE_NAME, {
      level: sanctumConfig.logLevel
    });
    addPlugin(resolver.resolve("./runtime/plugin"), { append: sanctumConfig.appendPlugin });
    addImportsDir(resolver.resolve("./runtime/composables"));
    if (sanctumConfig.globalMiddleware.enabled) {
      addRouteMiddleware({
        name: "sanctum:auth:global",
        path: resolver.resolve("./runtime/middleware/sanctum.global"),
        global: true
      }, {
        prepend: sanctumConfig.globalMiddleware.prepend
      });
      logger.info("Sanctum module initialized with global middleware");
    } else {
      addRouteMiddleware({
        name: "sanctum:auth",
        path: resolver.resolve("./runtime/middleware/sanctum.auth")
      });
      addRouteMiddleware({
        name: "sanctum:guest",
        path: resolver.resolve("./runtime/middleware/sanctum.guest")
      });
      logger.info("Sanctum module initialized w/o global middleware");
    }
    if (sanctumConfig.serverProxy.enabled) {
      addServerHandler({
        route: `${sanctumConfig.serverProxy.route}/**`,
        handler: resolver.resolve("./runtime/server/api/proxy")
      });
      logger.info("Sanctum module initialized with server proxy");
    }
    registerTypeTemplates(resolver);
  }
});

export { module as default };
