// nuxt-sanctum-precognition@0.0.9/dist/module.mjs
import { defineNuxtModule, createResolver, useLogger, installModule, addImportsDir } from '@nuxt/kit';
import defu from 'defu';

const defaultModuleOptions = {
  validateFiles: false,
  validationTimeout: 500,
  logLevel: 3
};

const MODULE_NAME = "nuxt-sanctum-precognition";
const module = defineNuxtModule({
  meta: {
    name: MODULE_NAME,
    configKey: "precognition"
  },
  defaults: defaultModuleOptions,
  async setup(_options, _nuxt) {
    const resolver = createResolver(import.meta.url);
    const precognitionConfig = defu(
      _nuxt.options.runtimeConfig.public.precognition,
      _options
    );
    _nuxt.options.build.transpile.push(resolver.resolve("./runtime"));
    _nuxt.options.runtimeConfig.public.precognition = precognitionConfig;
    const logger = useLogger(MODULE_NAME, {
      level: precognitionConfig.logLevel
    });
    await installModule("nuxt-auth-sanctum");
    addImportsDir(resolver.resolve("./runtime/composables"));
    logger.info("Precognition module initialized with Sanctum");
  }
});

export { module as default };
